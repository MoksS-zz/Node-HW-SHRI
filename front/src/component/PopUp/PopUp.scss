.PopUp {
  width: 100vw;
  height: 100vh;
  position: fixed;
  left: 0;
  top: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  background: rgba(0, 0, 0, 0.4);
}

.PopUp > * {
  background-color: var(--white);
}

.PopUp > .Form {
  padding: 7.5px 20px;
}