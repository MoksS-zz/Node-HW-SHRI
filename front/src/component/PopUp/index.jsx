import React, { useRef, useState } from "react";
import { useHistory } from "react-router-dom";
import Form from "./../Form";
import Input from "./../Input";
import Button from "./../Button";
import ButtonsField from "./../Button/ButtonsField";
import { host } from "../../helpers/constant";
import Text from "./../Text";
import "./PopUp.scss";

const PopUp = ({ hide }) => {
  const [error, setError] = useState("");
  const formEl = useRef(null);
  const history = useHistory();
  const onSaveClick = async (e) => {
    const bt1 = e.currentTarget;
    const bt2 = bt1.nextSibling;

    bt1.setAttribute("disabled", "disabled");
    bt2.setAttribute("disabled", "disabled");

    try {
      setError("");

      const formData = new FormData(formEl.current);

      const response = await fetch(`${host}/api/builds/${formData.get("commitHash")}`, {
        method: 'POST'
      });

      const result = await response.json();

      if(response.status >= 400) {
        setError(result.data);
        bt1.removeAttribute("disabled");
        bt2.removeAttribute("disabled");
        return;
      }

      history.push(`/build/${result.data.id}`);

    } catch (error) {
      setError(error);
      console.log(error);
    }

    bt1.removeAttribute("disabled");
    bt2.removeAttribute("disabled");
  };

  return (
    <div className="PopUp">
      <Form
        header="New Build"
        descripton="Enter the commit hash which you want to build."
        name="newBuild"
        method="post"
        formRef={formEl}
      >
        <Input
          type="text"
          placeholder="Commit Hash"
          name="commitHash"
        />

        <ButtonsField>
          <Button
            style={{ indentRigth: "s", indentBottom: "s", height: "action", color: "action", padding: "action" }}
            onClick={onSaveClick}
          >
            <Text style={{ size: "m", lineHeight: "xxxxl", weight: "small", color: "default" }}>Run build</Text>
          </Button>
          <Button
            onClick={hide}
            style={{ color: "close", padding: "action", height: "action", indentBottom: "xl" }}>
            <Text style={{ size: "m", lineHeight: "xxxxl", weight: "small", color: "default" }}>Cancel</Text>
          </Button>
        </ButtonsField>
        <Text style={{size: "l", weight: "normal", lineHeight: "xxl", color: "red"}}>{error}</Text>
      </Form>
    </div>
  )
}

export default PopUp;