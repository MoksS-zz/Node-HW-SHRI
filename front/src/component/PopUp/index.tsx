import React, { useRef, useState, FC, MouseEvent } from "react";
import { useHistory } from "react-router-dom";
import Form from "./../Form";
import Input from "./../Input";
import Button from "./../Button";
import ButtonsField from "./../Button/ButtonsField";
import { host } from "../../helpers/constant";
import Text from "./../Text";
import "./PopUp.scss";

export interface PopUpProps {
  hide(event: MouseEvent<HTMLButtonElement>): void;
}

const PopUp: FC<PopUpProps> = ({ hide }) => {
  const [error, setError] = useState("");
  const formEl = useRef<HTMLFormElement>(null);
  const history = useHistory();

  const { lang } = window;

  const onSaveClick= async (e: MouseEvent<HTMLButtonElement> ) => {
    const bt1 = e.currentTarget;
    const bt2 = bt1.nextSibling as HTMLButtonElement;

    bt1.setAttribute("disabled", "disabled");
    bt2.setAttribute("disabled", "disabled");

    try {
      setError("");

      const formData = new FormData(formEl.current!);

      const response = await fetch(`${host}/api/builds/${formData.get("commitHash")}`, {
        method: 'POST'
      });

      const result = await response.json();

      if(response.status >= 400) {
        setError(result.data);
        bt1.removeAttribute("disabled");
        bt2.removeAttribute("disabled");
        return;
      }

      history.push(`/build/${result.data.id}`);

    } catch (error) {
      setError(error);
      console.log(error);
    }

    bt1.removeAttribute("disabled");
    bt2.removeAttribute("disabled");
  };

  return (
    <div className="PopUp">
      <Form
        header={ lang.PopUp.Form.header }
        descripton={ lang.PopUp.Form.descripton }
        name="newBuild"
        method="post"
        formRef={formEl}
      >
        <Input
          type="text"
          placeholder={ lang.PopUp.Input.commitHash.placeholder }
          name="commitHash"
        />

        <ButtonsField>
          <Button
            style={{ indentRigth: "s", indentBottom: "s", height: "action", color: "action", padding: "action" }}
            onClick={onSaveClick}
          >
            <Text style={{ size: "m", lineHeight: "xxxxl", weight: "small", color: "default" }}>
            { lang.PopUp.Button.RunBuild }
            </Text>
          </Button>
          <Button
            onClick={hide}
            style={{ color: "close", padding: "action", height: "action", indentBottom: "xl" }}>
            <Text style={{ size: "m", lineHeight: "xxxxl", weight: "small", color: "default" }}>
            { lang.PopUp.Button.Cancel }
            </Text>
          </Button>
        </ButtonsField>
        <Text style={{size: "l", weight: "normal", lineHeight: "xxl", color: "red"}}>{error}</Text>
      </Form>
    </div>
  )
}

export default PopUp;