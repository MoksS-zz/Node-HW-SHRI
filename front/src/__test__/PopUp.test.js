import React from "react";
import renderer from "react-test-renderer";
import PopUp from "../component/PopUp";

jest.mock('react-router-dom', () => ({
  useHistory: jest.fn()
}))

it("renders PopUp correctly", () => {
  const logo = renderer
    .create(<PopUp hide={jest.fn()}/>)
    .toJSON();

  expect(logo).toMatchSnapshot();
});

