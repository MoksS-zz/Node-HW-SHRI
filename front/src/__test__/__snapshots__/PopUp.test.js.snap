// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PopUp correctly 1`] = `
<div
  className="PopUp"
>
  <form
    className="Form"
    method="post"
    name="newBuild"
    onClick={[Function]}
    onSubmit={[Function]}
  >
    <h2
      className="Form_H2"
    >
      New Build
    </h2>
    <p
      className="Form_Description"
    >
      Enter the commit hash which you want to build.
    </p>
    <div
      className="Input"
    >
      <p
        className="Input_Head"
      >
        
        
      </p>
      <div
        className="Input_Field"
      >
        <input
          className="Input_Input"
          name="commitHash"
          placeholder="Commit Hash"
          type="text"
        />
        <button
          className="Button Button__color-close"
          onClick={[Function]}
        >
          <span
            className="Icon Icon__size-xs Icon__img-close"
          />
        </button>
      </div>
      
    </div>
    <div
      className="ButtonsField"
    >
      <button
        className="Button Button__indentRigth-s Button__indentBottom-s Button__height-action Button__color-action Button__padding-action"
        onClick={[Function]}
      >
        <span
          className="Text Text__size-m Text__lineHeight-xxxxl Text__weight-small Text__color-default"
        >
          Run build
        </span>
      </button>
      <button
        className="Button Button__color-close Button__padding-action Button__height-action Button__indentBottom-xl"
        onClick={[MockFunction]}
      >
        <span
          className="Text Text__size-m Text__lineHeight-xxxxl Text__weight-small Text__color-default"
        >
          Cancel
        </span>
      </button>
    </div>
    <span
      className="Text Text__size-l Text__weight-normal Text__lineHeight-xxl Text__color-red"
    >
      
    </span>
  </form>
</div>
`;
